File: src/jimin/week6/순위 검색.py

# def solution(info, query):
#     answer = []
#     info_list = []
#     #언어, 직군, 경력, 소울 푸드, 점수
#     for i in info:
#         info_list.append(i.split(" "))
#
#     info_list = sorted(info_list, key= lambda x : int(x[4]))
#
#     query_list = []
#     for q in query:
#         q = q.replace("and", "")
#         q= q.split()
#         query_list.append(q)
#
#     n = len(info_list)
#     for q in query_list:
#         is_checked = [0] * n
#         for idx in range(5):
#             for info_idx, user_info in enumerate(info_list):
#                 if q[idx] == "-":
#                     continue
#                 if (idx == 4) and not is_checked[info_idx]:
#                     if int(q[idx]) > int(user_info[idx]):
#                         is_checked[info_idx] = 1
#                 elif q[idx] != user_info[idx] and not is_checked[info_idx]: #같지 않고, 이전까지 가능했다면
#                     is_checked[info_idx] = 1
#
#         answer.append(is_checked.count(0))
#     return answer
#
#
# print(solution(info,query))

from collections import defaultdict
from itertools import combinations

def binary_search(target, list):
    left, right = 0, len(list)
    while left < right:
        mid = (left + right) // 2
        if list[mid] < target:
            left = mid + 1
        else:
            right = mid
    return left

def solution(info, query):
    answer = []
    info_list = defaultdict(list)
    #key: 가능한 단어 조합, value: 수
    for i in info:
        for j in range(5):
            user_info = i.split(" ")
            for co in combinations(list(user_info[:4]), j):
                info_list["".join(co)].append(int(user_info[4]))
    print(info_list)

    for value in info_list.values():
        value.sort()

    for q in query:
        q = q.replace("and", "").replace("-","")
        q = q.split()
        count = int(q[-1])
        info_result = info_list["".join(q[:-1])]
        answer.append(len(info_result) - binary_search(count, info_result))
    return answer

info = ["java backend junior pizza 150","python frontend senior chicken 210","python frontend senior chicken 150","cpp backend senior pizza 260","java backend junior chicken 80","python backend senior chicken 50"]
query = ["java and backend and junior and pizza 100","python and frontend and senior and chicken 200","cpp and - and senior and pizza 250","- and backend and senior and - 150","- and - and - and chicken 100","- and - and - and - 150"]
print(solution(info,query))

File: src/taeeun/week6/순위검색.js
function solution(info, query) {
    var answer = [];
    const infoObj={};

    const binarySearch=(arr, target)=>{
        let left=0;
        let right=arr.length;
        while (left<right) {
            let mid=Math.floor((left+right)/2);
            if (arr[mid]>=target) right=mid;
            else left=mid+1;
        }
        return arr.length-left;
    }

    const combination=(info, score, infoObj, start)=>{
        const key=info.join(""); // 'javabackendjuniorpizza'

        //score을 배열 형태로 선언
        if (infoObj[key]) infoObj[key].push(score); //'javabackendjuniorpizza': [150, 170]
        else infoObj[key]=[score];

        for (let i=start; i<info.length; i++) {
            let temp=[...info];
            temp[i]='-';
            combination(temp, score, infoObj, i+1);
        }
    }

    for (let i=0; i<info.length; i++) {
        const infoArr=info[i].split(" "); //[ 'java', 'backend', 'junior', 'pizza']
        const score=Number(infoArr.pop());
        combination(infoArr, score, infoObj, 0);
    }


    const infoKeys=Object.keys(infoObj);

    const getResult=(query, score)=>{
        return infoKeys
        .filter(key=>query.every(e=>key.includes(e)))
        .reduce((acc, key) => { 
            return acc + binarySearch(infoObj[key], score)
        }, 0); 
    }

    for (let key in infoObj) {
        infoObj[key].sort((a,b)=>a-b);
    }    


    for (let i=0; i<query.length; i++){
        let newQuery=query[i].replace(/ and /g, "").split(" ");
        const score=Number(newQuery.pop());
        //score 점수 이상이면서 query 값을 만족해야 함
        answer.push(getResult(newQuery, score));
    }

    return answer;
}
